File: NaturaSwift/Classes/ASSharedTimeFormatter.m
//
//  ASSharedTimeFormatter.m
//  Naturapass
//
//  Created by Clément Padovani on 9/27/14.
//  Copyright (c) 2014 OCSMobi - 11. All rights reserved.
//

#import "ASSharedTimeFormatter.h"

static ASSharedTimeFormatter *_sharedFormatter = nil;

@interface ASSharedTimeFormatter ()

@property (nonatomic, strong, readwrite) NSDateFormatter *inputFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *outputFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *dateFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *timeFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *outputFormatterBis;

@property (nonatomic, strong, readwrite) NSDateFormatter *dateFormatterBis;

@property (nonatomic, strong, readwrite) NSDateFormatter *dateFormatterTer;

@property (nonatomic, strong, readwrite) NSDateFormatter *monthFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *weekDayFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *sectionHeaderViewDateFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *sectionHeaderViewDateRelativeFormatter;

@property (nonatomic, strong, readwrite) NSDateFormatter *timeFormatterBis;

@end

@implementation ASSharedTimeFormatter

+ (instancetype) sharedFormatter
{
	NSAssert([NSThread isMainThread], @"Not main thread! Thread: %@", [NSThread currentThread]);
	
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		_sharedFormatter = [[self alloc] initInStyle];
	});
	
	return _sharedFormatter;
}

- (instancetype) initInStyle
{
	self = [super init];
	
	if (self)
	{
		
	}
	
	return self;
}

- (instancetype) init
{
	NSAssert(NO, @"DON'T CALL ME!");
	
	self = [super init];
	
	if (self)
	{
		
	}
	
	return self;
}

// TODO: Ask Mathieu about lazy-init; if it's better to set the the ivar or the property

- (NSDateFormatter *) inputFormatter
{
	if (!_inputFormatter)
	{
		NSDateFormatter *inputFormatter = [[NSDateFormatter alloc] init];
		
		[inputFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[inputFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[inputFormatter setDateFormat: @"yyyy-MM-dd'T'HH:mm:ssZZZZ"];

		_inputFormatter = inputFormatter;
	}
	
	return _inputFormatter;
}

- (NSDateFormatter *) outputFormatter
{
	if (!_outputFormatter)
	{
		NSDateFormatter *outputFormatter = [[NSDateFormatter alloc] init];
		
		[outputFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[outputFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[outputFormatter setDateFormat: @"dd-MM-yyyy HH:mm"];
		
		_outputFormatter = outputFormatter;
	}
	
	return _outputFormatter;
}

- (NSDateFormatter *) dateFormatter
{
	if (!_dateFormatter)
	{
		NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
		
		[dateFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[dateFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[dateFormatter setDateFormat: @"dd/MM/yyyy"];
		
		_dateFormatter = dateFormatter;
	}
	
	return _dateFormatter;
}

- (NSDateFormatter *) timeFormatter
{
	if (!_timeFormatter)
	{
		NSDateFormatter *timeFormatter = [[NSDateFormatter alloc] init];
		
		[timeFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[timeFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[timeFormatter setDateFormat: @"dd-MM-yyyy H:mm"];

		_timeFormatter = timeFormatter;
	}
	
	return _timeFormatter;
}

- (NSDateFormatter *) outputFormatterBis
{
	if (!_outputFormatterBis)
	{
		NSDateFormatter *outputFormatterBis = [[NSDateFormatter alloc] init];
		
		[outputFormatterBis setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[outputFormatterBis setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[outputFormatterBis setDateFormat: @"d MMM yyyy à H:mm"];

		_outputFormatterBis = outputFormatterBis;
	}
	
	return _outputFormatterBis;
}

- (NSDateFormatter *) dateFormatterBis
{
	if (!_dateFormatterBis)
	{
		NSDateFormatter *dateFormatterBis = [[NSDateFormatter alloc] init];
		
		[dateFormatterBis setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[dateFormatterBis setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[dateFormatterBis setDateFormat: @"d MMM yyy"];
		
		_dateFormatterBis = dateFormatterBis;
	}
	
	return _dateFormatterBis;
}

- (NSDateFormatter *) dateFormatterTer
{
	if (!_dateFormatterTer)
	{
		NSDateFormatter *dateFormatterTer = [[NSDateFormatter alloc] init];
		
		[dateFormatterTer setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[dateFormatterTer setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[dateFormatterTer setDateFormat: @"dd/MM/yyyy HH:mm"];
		
		_dateFormatterTer = dateFormatterTer;
	}
	
	return _dateFormatterTer;
}

- (NSDateFormatter *) monthFormatter
{
	if (!_monthFormatter)
	{
		NSDateFormatter *monthFormatter = [[NSDateFormatter alloc] init];
		
		[monthFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[monthFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[monthFormatter setDateFormat: @"d MMMM"];

		_monthFormatter = monthFormatter;
	}
	
	return _monthFormatter;
}

- (NSDateFormatter *) weekDayFormatter
{
	if (!_weekDayFormatter)
	{
		NSDateFormatter *weekDayFormatter = [[NSDateFormatter alloc] init];
		
		[weekDayFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[weekDayFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[weekDayFormatter setDateFormat: @"EEEE"];
		
		_weekDayFormatter = weekDayFormatter;
	}
	
	return _weekDayFormatter;
}

- (NSDateFormatter *) sectionHeaderViewDateFormatter
{
	if (!_sectionHeaderViewDateFormatter)
	{
		NSDateFormatter *sectionHeaderViewDateFormatter = [[NSDateFormatter alloc] init];
		
		[sectionHeaderViewDateFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[sectionHeaderViewDateFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[sectionHeaderViewDateFormatter setDateFormat: @"EEEE d MMMM"];
		
		_sectionHeaderViewDateFormatter = sectionHeaderViewDateFormatter;
	}
	
	return _sectionHeaderViewDateFormatter;
}

- (NSDateFormatter *) sectionHeaderViewDateRelativeFormatter
{
	if (!_sectionHeaderViewDateRelativeFormatter)
	{
		NSDateFormatter *sectionHeaderViewDateRelativeFormatter = [[NSDateFormatter alloc] init];
		
		[sectionHeaderViewDateRelativeFormatter setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[sectionHeaderViewDateRelativeFormatter setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[sectionHeaderViewDateRelativeFormatter setTimeStyle: NSDateFormatterNoStyle];
		
		[sectionHeaderViewDateRelativeFormatter setDateStyle: NSDateFormatterShortStyle];
		
		[sectionHeaderViewDateRelativeFormatter setDoesRelativeDateFormatting: YES];
		
		_sectionHeaderViewDateRelativeFormatter = sectionHeaderViewDateRelativeFormatter;
	}
	
	return _sectionHeaderViewDateRelativeFormatter;
}

- (NSDateFormatter *) timeFormatterBis
{
	if (!_timeFormatterBis)
	{
		NSDateFormatter *timeFormatterBis = [[NSDateFormatter alloc] init];
		
		[timeFormatterBis setLocale: [NSLocale localeWithLocaleIdentifier: @"fr_FR"]];
		
		[timeFormatterBis setTimeZone: [NSTimeZone timeZoneWithName: @"Europe/Paris"]];
		
		[timeFormatterBis setDateFormat: @"HH:mm"];
		
		_timeFormatterBis = timeFormatterBis;
	}
	
	return _timeFormatterBis;
}

+ (void) checkFormatString: (NSString *) formatString forFormatter: (NSDateFormatter *) dateFormatter
{
	NSString *formatterFormatString = [dateFormatter dateFormat];
	
	NSAssert([formatString isEqualToString: formatterFormatString], @"ASSERT. %@ format string is supposed to be equal to %@", formatterFormatString, formatString);
}

@end

File: NaturaSwift/Classes/ASSharedTimeFormatter.swift
//
//  ASSharedTimeFormatter.swift
//  NaturaSwift
//
//  Created by Manh on 6/27/16.
//  Copyright © 2016 PHS. All rights reserved.
//

import Foundation
class ASSharedTimeFormatter: NSObject {
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: yyyy-MM-dd'T'HH:mm:ssZZZZ
     */
    
    var _inputFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: dd-MM-yyyy HH:mm
     */
    
    var _outputFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: dd/MM/yyyy
     */
    
    var _dateFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: dd-MM-yyyy H:mm
     */
    
    var _timeFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: d MMM yyyy à h:mm
     */
    
    var _outputFormatterBis:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: yyyy-M-d
     */
    
    var _dateFormatterBis:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: dd/MM/yyyy HH:mm
     */
    
    var _dateFormatterTer:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: d MMMM
     */
    
    var _monthFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  NSDateFormatter with date format of type: EEEE
     */
    
    var _weekDayFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-30 08:09
     *
     *	@brief  NSDateFormatter with date format of type: EEEE d MMMM
     */
    
    var _sectionHeaderViewDateFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-10-01 08:10
     *
     *	@brief  NSDateFormatter with relative formatting.
     */
    
    var _sectionHeaderViewDateRelativeFormatter:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-30 11:09
     *
     *	@brief  NSDateFormatter with date format of type: HH:mm
     */
    
    var _timeFormatterBis:NSDateFormatter!;
    
    /**
     *	@Author Clément Padovani, 14-09-29 10:09
     *
     *	@brief  Use this singleton to access the date formatters.
     *
     *	@return The shared time formatter singleton.
     */
    static func checkFormatString(formatString: String,forFormatter dateFormatter: NSDateFormatter)
    {
        let formatterFormatString: String = dateFormatter.dateFormat;
        assert(formatString == formatterFormatString, "ASSERT. \(formatterFormatString) format string is supposed to be equal to \(formatString)");
    }

    static let sharedFormatter = ASSharedTimeFormatter();
    
    override init() {
        super.init();
        
    }
    
    func inputFormatter() -> NSDateFormatter {
        if (_inputFormatter == nil) {
            let inputFormatter: NSDateFormatter = NSDateFormatter();
            inputFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            inputFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            inputFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZZZZ";
            _inputFormatter = inputFormatter;
        }
        return _inputFormatter;
    }
    func outputFormatter() -> NSDateFormatter {
        if (_outputFormatter == nil) {
            let outputFormatter: NSDateFormatter = NSDateFormatter();
            outputFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            outputFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            outputFormatter.dateFormat = "dd-MM-yyyy HH:mm";
            _outputFormatter = outputFormatter;
        }
        return _outputFormatter;
    }
    func dateFormatter() -> NSDateFormatter {
        if (_dateFormatter == nil) {
            let dateFormatter: NSDateFormatter = NSDateFormatter();
            dateFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            dateFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            dateFormatter.dateFormat = "dd/MM/yyyy";
            _dateFormatter = dateFormatter;
        }
        return _dateFormatter;
    }
    func timeFormatter() -> NSDateFormatter {
        if (_timeFormatter == nil) {
            let timeFormatter: NSDateFormatter = NSDateFormatter();
            timeFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            timeFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            timeFormatter.dateFormat = "dd-MM-yyyy H:mm";
            _timeFormatter = timeFormatter;
        }
        return _timeFormatter;
    }
    func outputFormatterBis() -> NSDateFormatter {
        if (_outputFormatterBis == nil) {
            let outputFormatterBis: NSDateFormatter = NSDateFormatter();
            outputFormatterBis.locale = NSLocale(localeIdentifier: "fr_FR");
            outputFormatterBis.timeZone = NSTimeZone(name: "Europe/Paris");
            outputFormatterBis.dateFormat = "d MMM yyyy à H:mm";
            _outputFormatterBis = outputFormatterBis;
        }
        return _outputFormatterBis;
    }
    func dateFormatterBis() -> NSDateFormatter {
        if (_dateFormatterBis == nil) {
            let dateFormatterBis: NSDateFormatter = NSDateFormatter();
            dateFormatterBis.locale = NSLocale(localeIdentifier: "fr_FR");
            dateFormatterBis.timeZone = NSTimeZone(name: "Europe/Paris");
            dateFormatterBis.dateFormat = "d MMM yyy";
            _dateFormatterBis = dateFormatterBis;
        }
        return _dateFormatterBis;
    }
    func dateFormatterTer() -> NSDateFormatter {
        if (_dateFormatterTer == nil) {
            let dateFormatterTer: NSDateFormatter = NSDateFormatter();
            dateFormatterTer.locale = NSLocale(localeIdentifier: "fr_FR");
            dateFormatterTer.timeZone = NSTimeZone(name: "Europe/Paris");
            dateFormatterTer.dateFormat = "dd/MM/yyyy HH:mm";
            _dateFormatterTer = dateFormatterTer;
        }
        return _dateFormatterTer;
    }
    func monthFormatter() -> NSDateFormatter {
        if (_monthFormatter == nil) {
            let monthFormatter: NSDateFormatter = NSDateFormatter();
            monthFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            monthFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            monthFormatter.dateFormat = "d MMMM";
            _monthFormatter = monthFormatter;
        }
        return _monthFormatter;
    }
    func fnweekDayFormatter() -> NSDateFormatter {
        if (_weekDayFormatter == nil) {
            let weekDayFormatter: NSDateFormatter = NSDateFormatter();
            weekDayFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            weekDayFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            weekDayFormatter.dateFormat = "d MMMM";
            _weekDayFormatter = weekDayFormatter;
        }
        return _weekDayFormatter;
    }
    func sectionHeaderViewDateFormatter() -> NSDateFormatter {
        if (_sectionHeaderViewDateFormatter == nil) {
            let sectionHeaderViewDateFormatter: NSDateFormatter = NSDateFormatter();
            sectionHeaderViewDateFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            sectionHeaderViewDateFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            sectionHeaderViewDateFormatter.dateFormat = "EEEE d MMMM";
            _sectionHeaderViewDateFormatter = sectionHeaderViewDateFormatter;
        }
        return _sectionHeaderViewDateFormatter;
    }
    func sectionHeaderViewDateRelativeFormatter() -> NSDateFormatter {
        if (_sectionHeaderViewDateRelativeFormatter == nil) {
            let sectionHeaderViewDateRelativeFormatter: NSDateFormatter = NSDateFormatter();
            sectionHeaderViewDateRelativeFormatter.locale = NSLocale(localeIdentifier: "fr_FR");
            sectionHeaderViewDateRelativeFormatter.timeZone = NSTimeZone(name: "Europe/Paris");
            sectionHeaderViewDateRelativeFormatter.timeStyle = NSDateFormatterStyle.NoStyle;
            sectionHeaderViewDateRelativeFormatter.dateStyle = NSDateFormatterStyle.ShortStyle;
            sectionHeaderViewDateRelativeFormatter.doesRelativeDateFormatting = true;
            _sectionHeaderViewDateRelativeFormatter = sectionHeaderViewDateRelativeFormatter;
        }
        return _sectionHeaderViewDateRelativeFormatter;
    }
    func timeFormatterBis() -> NSDateFormatter {
        if (_timeFormatterBis == nil) {
            let timeFormatterBis: NSDateFormatter = NSDateFormatter();
            timeFormatterBis.locale = NSLocale(localeIdentifier: "fr_FR");
            timeFormatterBis.timeZone = NSTimeZone(name: "Europe/Paris");
            timeFormatterBis.dateFormat = "HH:mm";
            _timeFormatterBis = timeFormatterBis;
        }
        return _timeFormatterBis;
    }
    
    
}